### DISCUSSION TITLE
Support foreach loop with ref value as loop value for arrays, spans and other sequence classes

### DISCUSSION THREAD

**Discussion by AlexRadch** [dotnet/csharplang#6910] [Language Ideas]
The `foreach` loop is handy for processing all values in sequence.  But it does not support changing sequence values.
When we need to change sequence values, we must use other ugly loops (`for`, `while`, etc.).
I suggest adding support of ref value in the `foreach` loop.

```cs
var lists = new List<string>[N]; // array of string lists
foreach (ref var list in lists) // ref to List<string>
    list = new List<string>(); // Initialize all lists without ugly for loop

var items = new int[N]; // or new List<int>(); or other sequence with write access to items.
foreach (ref var item in items)
    item = item * 2; // Sometimes it will be two times faster then use items[i] = items[i] * 2.

var dict = new Dictionary<Key, Value>();
foreach (ref var value in dict.Values)
    value = Process(value); // Process dictionary values without keys and CollectionsMarshal using
```

It will increase performance for sequences with large value types also.

To support the `for each` loop with ref value at any sequence class, I suggest adding interfaces `IRefEnumerable` for sequences with write access and `IRefInEnumerable` for sequences with read access. Duck typing should be supported also.

**Comment by svick** [comment]
That feature already exists in C#. Though it has a couple of problems:

1. I don't think any type in the BCL supports it, so to use it, you'll have to write your own extensions.
2. Specifically for array, extension `GetEnumerator()` doesn't work (https://github.com/dotnet/roslyn/issues/44623), so you'll have to use uglier syntax.

To make it concrete, [the following code works now](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4AEAMACFBGALAbgFgAoEgYwHsA7AZ2AxivoDkMBeDAkksAQygwAbGHRrsMVCAHcMAGRHAAPDjQA+ANrMAukWJwKUCLzIALDAApDcDHwHC6DKkIU0AdAHEIwAEoQ4AUSoAVwBbaF4AI0EIcwBKWJIMJOcHDkkZeTplbDU43RIUAGYMQ0ooABMMOiggsnoAQSgoXgRA0PDgA0UAFVVzbvUtDH5mhATiAG9E5MZ6RnKIAA9xAFpsXWSsYqsMbowAYSCmiCZ2VRK/YaaW9XmlnWmkoowIigpBDABZCkhmJeA4mcMABqYF3ZaKK6jVyyE4Ac2AJl0AF9uMRnqUDJVqrUGtdWsEws0ohAen0BkMRi1xlNiJtno1Rm0ibxOlAyRhPMBmR0DIC2Od0uYqWMUWicAA2LAAJgw/kWwBONBg1BoJFp9OKkowjJaPOJ0Q5XN8AUJ4RJZPMiJEu0GUOpQKFItiKKAA==):

```c#
using System.Collections.Generic;

const int N = 4;

var lists = new List<string>[N];
foreach (ref var list in lists.GetRefEnumerable())
    list = new List<string>();

public record struct ArrayEnumerator<T>(T[] array)
{
    int index = -1;
    public ref T Current => ref array[index];
    public bool MoveNext() => ++index < array.Length;
}

public record struct ArrayEnumerable<T>(T[] array)
{
    public ArrayEnumerator<T> GetEnumerator() => new(array);
}

static class Extensions
{
    public static ArrayEnumerable<T> GetRefEnumerable<T>(this T[] array) => new(array);
}
```

**Reply by AlexRadch** [reply]
Thanks for the second part of the answer, it is very clean.

But I did not understand issues with BCL.
Can we add the new interface to the Array class to support the suggested syntax?
Can we add new methods to the Array class to support the suggested syntax as duck typing?

**Reply by TahirAhmadov** [reply]
I think BCL cannot add this feature at least until the referenced Roslyn bug is resolved.

**Comment by AlexRadch** [comment]
CharGPT approves my proposal 🤣🤣🤣:
![изображение](https://user-images.githubusercontent.com/6407980/214390326-df6a5817-7240-41f7-af5c-2cacd64cc5f5.png)

**Comment by bernd5** [comment]
Today you can write:
```csharp
using System;

int[] array = new[]{1, 2, 3};
foreach(ref var item in array.AsSpan()){
    item *= 2;
}
```
The "problem" with the other BCL collection types is that they do not offer a by-ref access. I assume this is because of backward compatibility.
In addition they do not garantuee sequential memory - they can't be wrapped / converted to a span... (List could, but e.g. access after adding could not be detected...)

Perhaps you should open an issue in the runtime repo...

**Reply by tannergooding** [reply]
They do not offer byref access because it is unsafe to do so.

We do have unsafe ways to get a ref to `Dictionary<TKey, TValue>` and a span to `List<T>`, but they are hidden due to their unsafe nature.

**Reply by AlexRadch** [reply]
> Today you can write: `foreach(ref var item in array.AsSpan())`

One more argument to support my suggestion as a standard way to iterate items on writable sequences. And `foreach(ref readonly var item in sequence)` on readonly sequences (ReadonlyList and etc.).

**Reply by AlexRadch** [reply]
> They do not offer byref access because it is unsafe to do so.
> 
> We do have unsafe ways to get a ref to `Dictionary<TKey, TValue>` and a span to `List<T>`, but they are hidden due to their unsafe nature.

The `CollectionsMarshal` is unsafe but we can write a safe `IRefEnumerable` implementation for dictionary values and a safe `IRefReadonlyEnumerable` implementation for dictionary keys. And they will have more performance than standard enumerators in some cases.

**Comment by huoyaoyuan** [comment]
Related: #461
